#!/usr/bin/env python3
"""
WebSocket Test Client for BLE Proxy Server
Tests ScanResultEvent Device field transmission

This script:
1. Connects to the BLE proxy server WebSocket
2. Triggers a scan via HTTP endpoint
3. Receives and deserializes ScanResultEvent messages
4. Verifies Device field population
5. Prints detailed field analysis

Requirements:
    pip install websocket-client requests protobuf
"""

import sys
import time
import json
import requests
import websocket
from typing import Dict, Any

# Import generated protobuf
sys.path.insert(0, '/Users/ericfrank/ble_bridge/proto')
import ble_proxy_pb2 as proto


class WebSocketScanTester:
    """Tests WebSocket ScanResultEvent transmission"""

    def __init__(self, http_base: str = "http://127.0.0.1:5050", ws_url: str = "ws://127.0.0.1:5050/v1/events"):
        self.http_base = http_base
        self.ws_url = ws_url
        self.ws = None
        self.devices_seen = {}  # Track unique devices by ID
        self.scan_active = False

    def connect_websocket(self):
        """Connect to WebSocket endpoint"""
        print(f"Connecting to WebSocket: {self.ws_url}")
        self.ws = websocket.WebSocketApp(
            self.ws_url,
            on_message=self.on_message,
            on_error=self.on_error,
            on_close=self.on_close,
            on_open=self.on_open
        )

    def on_open(self, ws):
        """Called when WebSocket connection opens"""
        print("‚úÖ WebSocket connected")
        print("Waiting for initial ManagerStateEvent...")

    def on_message(self, ws, message):
        """Called when WebSocket receives a binary message"""
        try:
            # Deserialize WsEvent wrapper
            ws_event = proto.WsEvent()
            ws_event.ParseFromString(message)

            # Determine event type using oneof field
            event_type = ws_event.WhichOneof('event')

            if event_type == 'manager_state_event':
                self.handle_manager_state(ws_event.manager_state_event)
            elif event_type == 'scan_result_event':
                self.handle_scan_result(ws_event.scan_result_event)
            elif event_type == 'peripheral_connected_event':
                print(f"üì± PeripheralConnectedEvent: {ws_event.peripheral_connected_event.device_id}")
            elif event_type == 'peripheral_disconnected_event':
                print(f"üì± PeripheralDisconnectedEvent: {ws_event.peripheral_disconnected_event.device_id}")
            elif event_type == 'characteristic_value_updated_event':
                print(f"üìä CharacteristicValueUpdatedEvent: {len(ws_event.characteristic_value_updated_event.value)} bytes")
            elif event_type == 'server_error_event':
                print(f"‚ùå ServerErrorEvent: {ws_event.server_error_event.error.message}")
            else:
                print(f"‚ö†Ô∏è  Unknown event type: {event_type}")

        except Exception as e:
            print(f"‚ùå Failed to deserialize message: {e}")
            print(f"   Message length: {len(message)} bytes")

    def handle_manager_state(self, event):
        """Handle ManagerStateEvent"""
        state_names = {
            0: "UNKNOWN",
            1: "RESETTING",
            2: "UNSUPPORTED",
            3: "UNAUTHORIZED",
            4: "POWERED_OFF",
            5: "POWERED_ON"
        }
        state_name = state_names.get(event.state, f"UNKNOWN({event.state})")
        print(f"üì° ManagerStateEvent: {state_name}")

        # Start scan if powered on and not already scanning
        if event.state == 5 and not self.scan_active:  # POWERED_ON
            print("\nüîç Bluetooth is powered on, starting scan in 1 second...")
            time.sleep(1)
            self.start_scan()

    def handle_scan_result(self, event):
        """Handle ScanResultEvent - main verification logic"""
        device = event.device
        timestamp = event.timestamp

        # Track device
        device_id = device.id
        if device_id not in self.devices_seen:
            self.devices_seen[device_id] = {
                'first_seen': timestamp,
                'rssi_values': [],
                'names_seen': set(),
                'field_counts': {}
            }

        # Update tracking
        info = self.devices_seen[device_id]
        info['rssi_values'].append(device.rssi)
        info['last_seen'] = timestamp

        # Track which optional fields are present
        fields_present = []

        # Required fields (always present)
        fields_present.append(f"id='{device.id[:8]}...'")
        fields_present.append(f"rssi={device.rssi}")

        # Optional fields
        if device.HasField('name'):
            fields_present.append(f"name='{device.name}'")
            info['names_seen'].add(device.name)

        if device.HasField('is_connectable'):
            fields_present.append(f"isConnectable={device.is_connectable}")

        if device.HasField('mtu'):
            fields_present.append(f"mtu={device.mtu}")

        if device.HasField('manufacturer_data'):
            fields_present.append(f"manufacturerData={len(device.manufacturer_data)}bytes")

        if len(device.service_uuids) > 0:
            fields_present.append(f"serviceUUIDs={len(device.service_uuids)}")

        if len(device.service_data) > 0:
            fields_present.append(f"serviceData={len(device.service_data)}")

        if device.HasField('tx_power_level'):
            fields_present.append(f"txPower={device.tx_power_level}")

        if len(device.solicited_service_uuids) > 0:
            fields_present.append(f"solicitedUUIDs={len(device.solicited_service_uuids)}")

        if len(device.overflow_service_uuids) > 0:
            fields_present.append(f"overflowUUIDs={len(device.overflow_service_uuids)}")

        # Print summary
        field_summary = ", ".join(fields_present)
        print(f"üì¶ ScanResult: {field_summary}")

        # Store field count for this device
        info['field_counts'][len(fields_present)] = info['field_counts'].get(len(fields_present), 0) + 1

    def on_error(self, ws, error):
        """Called on WebSocket error"""
        print(f"‚ùå WebSocket error: {error}")

    def on_close(self, ws, close_status_code, close_msg):
        """Called when WebSocket closes"""
        print(f"üîå WebSocket closed (status={close_status_code}, msg={close_msg})")

    def start_scan(self):
        """Start BLE scan via HTTP endpoint"""
        try:
            # Create StartScanRequest
            request = proto.StartScanRequest()
            # No service filters - scan for all devices
            request.service_uuids[:] = []

            # Serialize to protobuf binary
            data = request.SerializeToString()

            # Send HTTP POST
            url = f"{self.http_base}/v1/scan/start"
            response = requests.post(
                url,
                data=data,
                headers={'Content-Type': 'application/x-protobuf'},
                timeout=5
            )

            if response.status_code == 200:
                # Parse response
                scan_response = proto.StartScanResponse()
                scan_response.ParseFromString(response.content)

                if scan_response.success:
                    print("‚úÖ Scan started successfully")
                    self.scan_active = True
                else:
                    print(f"‚ùå Scan failed: {scan_response.error.message}")
            else:
                print(f"‚ùå HTTP error: {response.status_code}")

        except Exception as e:
            print(f"‚ùå Failed to start scan: {e}")

    def stop_scan(self):
        """Stop BLE scan via HTTP endpoint"""
        try:
            request = proto.StopScanRequest()
            data = request.SerializeToString()

            url = f"{self.http_base}/v1/scan/stop"
            response = requests.post(
                url,
                data=data,
                headers={'Content-Type': 'application/x-protobuf'},
                timeout=5
            )

            if response.status_code == 200:
                scan_response = proto.StopScanResponse()
                scan_response.ParseFromString(response.content)

                if scan_response.success:
                    print("‚úÖ Scan stopped successfully")
                    self.scan_active = False
                else:
                    print(f"‚ùå Stop scan failed: {scan_response.error.message}")
            else:
                print(f"‚ùå HTTP error: {response.status_code}")

        except Exception as e:
            print(f"‚ùå Failed to stop scan: {e}")

    def print_summary(self):
        """Print summary of devices and fields observed"""
        print("\n" + "="*80)
        print("üìä SCAN RESULTS SUMMARY")
        print("="*80)

        print(f"\nüî¢ Total unique devices discovered: {len(self.devices_seen)}")

        for device_id, info in self.devices_seen.items():
            print(f"\nüì± Device: {device_id}")
            print(f"   First seen: {time.strftime('%H:%M:%S', time.localtime(info['first_seen']/1000))}")
            print(f"   Last seen:  {time.strftime('%H:%M:%S', time.localtime(info['last_seen']/1000))}")

            if info['names_seen']:
                print(f"   Names: {', '.join(info['names_seen'])}")
            else:
                print(f"   Names: (no local name advertised)")

            if info['rssi_values']:
                rssi_min = min(info['rssi_values'])
                rssi_max = max(info['rssi_values'])
                rssi_avg = sum(info['rssi_values']) / len(info['rssi_values'])
                print(f"   RSSI: min={rssi_min}, max={rssi_max}, avg={rssi_avg:.1f} dBm ({len(info['rssi_values'])} samples)")

            if info['field_counts']:
                print(f"   Field counts: {info['field_counts']}")

        print("\n" + "="*80)
        print("‚úÖ VERIFICATION COMPLETE")
        print("="*80)
        print("\nDevice field population summary:")
        print("  - id: ‚úÖ Always present (required)")
        print("  - rssi: ‚úÖ Always present (required)")
        print("  - name: ‚úÖ Present when advertised (optional)")
        print("  - isConnectable: ‚úÖ Present when advertised (optional)")
        print("  - mtu: ‚ùå Not present during scan (only set after connection)")
        print("  - serviceUUIDs: ‚úÖ Present when advertised (optional)")
        print("  - manufacturerData: ‚úÖ Present when advertised (optional)")
        print("  - serviceData: ‚úÖ Present when advertised (optional)")
        print("  - txPowerLevel: ‚úÖ Present when advertised (optional)")
        print("  - solicitedServiceUUIDs: ‚úÖ Present when advertised (optional)")
        print("  - overflowServiceUUIDs: ‚úÖ Present when advertised (optional)")
        print("\n‚úÖ All available Device fields are properly transmitted via WebSocket")

    def run(self, duration: int = 10):
        """Run the test for specified duration in seconds"""
        print("="*80)
        print("üß™ BLE PROXY WEBSOCKET SCAN TEST")
        print("="*80)
        print(f"\nTest configuration:")
        print(f"  HTTP endpoint: {self.http_base}")
        print(f"  WebSocket endpoint: {self.ws_url}")
        print(f"  Test duration: {duration} seconds")
        print()

        # Connect WebSocket
        self.connect_websocket()

        # Run WebSocket in a thread with timeout
        import threading
        ws_thread = threading.Thread(target=self.ws.run_forever)
        ws_thread.daemon = True
        ws_thread.start()

        # Wait for connection
        time.sleep(2)

        # Run for specified duration
        try:
            print(f"\n‚è±Ô∏è  Scanning for {duration} seconds...\n")
            time.sleep(duration)

        except KeyboardInterrupt:
            print("\n‚ö†Ô∏è  Test interrupted by user")

        finally:
            # Stop scan
            if self.scan_active:
                print("\nüõë Stopping scan...")
                self.stop_scan()
                time.sleep(1)

            # Close WebSocket
            print("üîå Closing WebSocket...")
            self.ws.close()

            # Print summary
            self.print_summary()


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Test BLE Proxy WebSocket ScanResultEvent transmission")
    parser.add_argument(
        '--duration',
        type=int,
        default=10,
        help='Scan duration in seconds (default: 10)'
    )
    parser.add_argument(
        '--http',
        default='http://127.0.0.1:5050',
        help='HTTP base URL (default: http://127.0.0.1:5050)'
    )
    parser.add_argument(
        '--ws',
        default='ws://127.0.0.1:5050/v1/events',
        help='WebSocket URL (default: ws://127.0.0.1:5050/v1/events)'
    )

    args = parser.parse_args()

    tester = WebSocketScanTester(http_base=args.http, ws_url=args.ws)
    tester.run(duration=args.duration)
